engine/consensus/approvals/tracker/record.go

```diff
@@ -1,66 +1,133 @@
 
-	// SufficientApprovalsForSealing: True iff all chunks in the result have
-	// sufficient approvals
-	SufficientApprovalsForSealing bool
+	// entries holds the individual entries of the sealing record
+	entries Rec
+}
 
-	// firstUnmatchedChunkIndex: Index of first chunk that hasn't received
-	// sufficient approval (ordered by chunk index). Optional value: only set
-	// if SufficientApprovalsForSealing == False and nil otherwise.
-	firstUnmatchedChunkIndex *uint64
+func (r *SealingRecord) QualifiesForEmergencySealing(emergencySealable bool) {
+	r.entries["qualifies_for_emergency_sealing"] = emergencySealable
+}
 
-	// qualifiesForEmergencySealing: True iff result qualifies for emergency
-	// sealing. Optional value: only set if
-	// SufficientApprovalsForSealing == False and nil otherwise.
-	qualifiesForEmergencySealing *bool
+func (r *SealingRecord) ApprovalsMissing(chunksWithMissingApprovals map[uint64]flow.IdentifierList) {
+	sufficientApprovals := len(chunksWithMissingApprovals) == 0
+	r.entries["sufficient_approvals_for_sealing"] = sufficientApprovals
+	if !sufficientApprovals {
+		indices := make([]uint64, 0, len(chunksWithMissingApprovals))
+		for i := range chunksWithMissingApprovals {
+			indices = append(indices, i)
+		}
+		r.entries["chunks_with_insufficient_approvals"] = indices
+	}
 
-	// hasMultipleReceipts: True iff there are at least 2 receipts from
-	// _different_ ENs committing to the result. Optional value: only set if
-	// SufficientApprovalsForSealing == True and nil otherwise.
-	hasMultipleReceipts *bool
 }
 
-// NewRecordWithSufficientApprovals creates a sealing record for an
-// incorporated result with sufficient approvals for sealing.
-func NewRecordWithSufficientApprovals(ir *flow.IncorporatedResult) *SealingRecord {
-	return &SealingRecord{
-		IncorporatedResult:            ir,
-		SufficientApprovalsForSealing: true,
-	}
+func (r *SealingRecord) ApprovalsRequested(requestCount uint) {
+	r.entries["number_requested_approvals"] = requestCount
 }
 
-// NewRecordWithInsufficientApprovals creates a sealing record for an
-// incorporated result that has insufficient approvals to be sealed.
-// firstUnmatchedChunkIndex specifies the index of first chunk that
-// hasn't received sufficient approval.
-func NewRecordWithInsufficientApprovals(ir *flow.IncorporatedResult, firstUnmatchedChunkIndex uint64) *SealingRecord {
-	return &SealingRecord{
-		IncorporatedResult:            ir,
-		SufficientApprovalsForSealing: false,
-		firstUnmatchedChunkIndex:      &firstUnmatchedChunkIndex,
+// Generate generates a key-value map capturing the application-submitted data
+// plus auxiliary data.
+func (r *SealingRecord) Generate() (Rec, error) {
+	rec := make(Rec)
+	for k, v := range r.entries {
+		rec[k] = v
+	}
+
+	irID := r.IncorporatedResult.ID()
+	result := r.IncorporatedResult.Result
+	resultID := result.ID()
+	executedBlock, err := r.headersDB.ByBlockID(result.BlockID)
+	if err != nil {
```

Comment:
We fail the whole map generation on the first failure to retrieve an output field. Do we have a hang of how often we expect retrieval to fail for some (but not all) fields?

Reply:
These errors should not happen _at all_. If they do, we have either bugs or a corrupted internal state (like a corrupted DB, cache, etc). 

When such error occurs, we log it as such: https://github.com/onflow/flow-go/blob/791cf86012ad429fac4ebe8f3ec13c7873d30b5d/engine/consensus/approvals/tracker/tracker.go#L153-L159